sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,306 @@ import org.apache.spark.util.Utils
+      // guard the non string type
+      null
+
+    case token =>
+    // We cannot parse this token based on the given data type. So, we throw a
+    // SparkSQLJsonProcessingException and this exception will be caught by
+    // parseJson method.
+    throw new SparkSQLJsonProcessingException(
+      s"Failed to parse a value for data type $dataType (current token: $token).")
+  }
+
+  private def failedRecord(record: String): Seq[InternalRow] = {
+    // create a row even if no corrupt record column is present
+    if (options.failFast) {
+      throw new RuntimeException(s"Malformed line in FAILFAST mode: $record")
+    }
+    if (options.dropMalformed) {
+      logWarning(s"Dropping malformed line: $record")
+      Nil
+    } else {
+      val row = new GenericMutableRow(schema.length)
+      for (corruptIndex <- schema.getFieldIndex(options.columnNameOfCorruptRecord)) {
+        require(schema(corruptIndex).dataType == StringType)
+        row.update(corruptIndex, UTF8String.fromString(record))
+      }
+      Seq(row)
     }
   }
 
   /**
-   * Parse the current token (and related children) according to a desired schema
-   * This is a wrapper for the method `convertField()` to handle a row wrapped
-   * with an array.
+   * Create a converter which converts the JSON documents held by the `JsonParser`
+   * to a value according to a desired schema. This is a wrapper for the method
+   * `makeConverter()` to handle a row wrapped with an array.
    */
-  def convertRootField(
-      factory: JsonFactory,
-      parser: JsonParser,
-      schema: DataType): Any = {
-    import com.fasterxml.jackson.core.JsonToken._
-    (parser.getCurrentToken, schema) match {
-      case (START_ARRAY, st: StructType) =>
-        // SPARK-3308: support reading top level JSON arrays and take every element
-        // in such an array as a row
-        convertArray(factory, parser, st)
-
-      case (START_OBJECT, ArrayType(st, _)) =>
-        // the business end of SPARK-3308:
-        // when an object is found but an array is requested just wrap it in a list
-        convertField(factory, parser, st) :: Nil
+  def makeRootConverter(dataType: DataType): ValueConverter = dataType match {
+    case st: StructType =>
+      // SPARK-3308: support reading top level JSON arrays and take every element
+      // in such an array as a row
+      val elementConverter = makeConverter(st)
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => parser.getCurrentToken match {
+        case START_OBJECT => convertObject(parser, st, fieldConverters)
+        case START_ARRAY => convertArray(parser, elementConverter)
+        case _ => failedConversion(parser, st)
+      }
 
-      case _ =>
-        convertField(factory, parser, schema)
-    }
+    case ArrayType(st: StructType, _) =>
+      // the business end of SPARK-3308:
+      // when an object is found but an array is requested just wrap it in a list
```

Comment:
also, this comment should be put at https://github.com/apache/spark/pull/14102/files#diff-8affe5ec7d691943a88e43eb30af656eR110, right?